codec-http/src/main/java/io/netty/handler/codec/http/HttpContentDecoder.java

```diff
@@ -73,76 +71,62 @@ protected void decode(ChannelHandlerContext ctx, HttpObject msg, List<Object> ou
-                        int contentLength = 0;
-                        int size = out.size();
-                        for (int i = 0; i < size; i++) {
-                            Object o = out.get(i);
-                            if (o instanceof HttpContent) {
-                                contentLength += ((HttpContent) o).content().readableBytes();
-                            }
-                        }
-                        headers.set(
-                                HttpHeaderNames.CONTENT_LENGTH,
-                                Integer.toString(contentLength));
-                    }
-                    return;
-                }
+            // Determine the content encoding.
+            String contentEncoding = headers.getAndConvert(HttpHeaderNames.CONTENT_ENCODING);
+            if (contentEncoding != null) {
+                contentEncoding = contentEncoding.trim();
+            } else {
+                contentEncoding = IDENTITY;
+            }
+            decoder = newContentDecoder(contentEncoding);
 
-                if (c instanceof LastHttpContent) {
-                    decodeStarted = false;
+            if (decoder == null) {
+                if (message instanceof HttpContent) {
+                    ((HttpContent) message).retain();
                 }
                 out.add(message);
-                out.add(c.retain());
                 return;
             }
 
-            if (decoder != null) {
-                decodeContent(c, out);
+            // Remove content-length header:
+            // the correct value can be set only after all chunks are processed/decoded.
+            // If buffering is not an issue, add HttpObjectAggregator down the chain, it will set the header.
+            // Otherwise, rely on LastHttpContent message.
+            headers.remove(HttpHeaderNames.CONTENT_LENGTH);
+
+            // set new content encoding,
+            CharSequence targetContentEncoding = getTargetContentEncoding(contentEncoding);
+            if (HttpHeaderValues.IDENTITY.equals(targetContentEncoding)) {
+                // Do NOT set the 'Content-Encoding' header if the target encoding is 'identity'
+                // as per: http://tools.ietf.org/html/rfc2616#section-14.11
+                headers.remove(HttpHeaderNames.CONTENT_ENCODING);
             } else {
-                if (c instanceof LastHttpContent) {
-                    decodeStarted = false;
-                }
+                headers.set(HttpHeaderNames.CONTENT_ENCODING, targetContentEncoding);
+            }
+
+            if (message instanceof HttpContent) {
+                // If message is a full request (headers + data), don't copy data part into out.
+                // Output headers only; data part will be decoded below
+                HttpRequest r = (HttpRequest) message;
+                HttpRequest copy = new DefaultHttpRequest(r.protocolVersion(), r.method(), r.uri());
```

Comment:
If the expectation is you are dealing with a `FullHttpRequest` then can you use [FullHttpRequest.copy(null)](https://github.com/netty/netty/blob/master/codec-http/src/main/java/io/netty/handler/codec/http/FullHttpRequest.java#L25)?  This will prevent duplicating the copy code and provide the same result.  We can discuss if `setDecoderResult` should be in the `copy` method.

The same method is available in [FullHttpResponse](https://github.com/netty/netty/blob/master/codec-http/src/main/java/io/netty/handler/codec/http/FullHttpResponse.java#L25)